exec/java-exec/src/main/java/org/apache/drill/exec/expr/stat/ParquetIsPredicate.java

```diff
@@ -62,50 +62,51 @@ private ParquetIsPredicate(LogicalExpression expr, BiPredicate<Statistics<C>, Ra
     return visitor.visitUnknown(this, value);
   }
 
-  @Override
-  public boolean canDrop(RangeExprEvaluator<C> evaluator) {
+  /**
+   * Apply the filter condition against the meta of the rowgroup.
+   */
+  public RowsMatch matches(RangeExprEvaluator<C> evaluator) {
     Statistics<C> exprStat = expr.accept(evaluator, null);
-    if (isNullOrEmpty(exprStat)) {
-      return false;
-    }
+    return isNullOrEmpty(exprStat) ? RowsMatch.SOME : predicate.apply(exprStat, evaluator);
+  }
 
-    return predicate.test(exprStat, evaluator);
+  /**
+   * After the applying of the filter against the statistics of the rowgroup, if the result is RowsMatch.ALL,
+   * then we still must know if the rowgroup contains some null values, because it can change the filter result.
+   * If it contains some null values, then we change the RowsMatch.ALL into RowsMatch.SOME, which sya that maybe
+   * some values (the null ones) should be disgarded.
+   */
+  private static RowsMatch checkNull(Statistics exprStat) {
+    return hasNoNulls(exprStat) ? RowsMatch.ALL : RowsMatch.SOME;
   }
 
   /**
    * IS NULL predicate.
    */
   private static <C extends Comparable<C>> LogicalExpression createIsNullPredicate(LogicalExpression expr) {
     return new ParquetIsPredicate<C>(expr,
-        //if there are no nulls  -> canDrop
-        (exprStat, evaluator) -> hasNoNulls(exprStat)) {
-      private final boolean isArray = isArray(expr);
-
-      private boolean isArray(LogicalExpression expression) {
-        if (expression instanceof TypedFieldExpr) {
-          TypedFieldExpr typedFieldExpr = (TypedFieldExpr) expression;
-          SchemaPath schemaPath = typedFieldExpr.getPath();
-          return schemaPath.isArray();
-        }
-        return false;
-      }
-
-      @Override
-      public boolean canDrop(RangeExprEvaluator<C> evaluator) {
+      (exprStat, evaluator) -> {
         // for arrays we are not able to define exact number of nulls
         // [1,2,3] vs [1,2] -> in second case 3 is absent and thus it's null but statistics shows no nulls
-        return !isArray && super.canDrop(evaluator);
-      }
-    };
+        if (expr instanceof TypedFieldExpr) {
+          TypedFieldExpr typedFieldExpr = (TypedFieldExpr) expr;
+          if (typedFieldExpr.getPath().isArray()) {
+            return RowsMatch.SOME;
+          }
+        }
+        if (hasNoNulls(exprStat)) {
```

Comment:
@jbimbert, @vrozov 
https://issues.apache.org/jira/browse/DRILL-6603 use case shows that statistics can be defined in the following way:
`stat.isEmpty()` returns `false`
`stat.getNumNulls()` returns `0`
`stat.genericGetMin()` returns `null`
`stat.genericGetMax()` returns `null`
`stat.toString` returns `num_nulls: 0, min/max not defined`
In this case we mistakenly decide that such row group does not contain null values and filter out such rows group but in reality, row group contains nulls.
Statistics `toString` method is defined as follows:
```
  @Override
  public String toString() {
    if (this.hasNonNullValue()) {
      if (isNumNullsSet()) {
        return String.format("min: %s, max: %s, num_nulls: %d", minAsString(), maxAsString(), this.getNumNulls());
      } else {
        return String.format("min: %s, max: %s, num_nulls not defined", minAsString(), maxAsString());
      }
    } else if (!this.isEmpty())
      return String.format("num_nulls: %d, min/max not defined", this.getNumNulls());
    else
      return "no stats for this column";
  }
```
As you can see `toString` method also uses `stat.hasNonNullValue()` to show if stat is defined.
In order to correctly apply `is null` we need use `stat.hasNonNullValue()`.

Reply:
In testBooleanPredicate, file java-exec/src/test/resources/parquetFilterPush/blnTbl/0_0_3.parquet which has ST:[num_nulls: 3, min/max not defined] and "row group 1: RC:3 TS:27 OFFSET:4", we get the correct value isAllNulls=true, because we compare getNumNulls with rowCount, which are both 3 in this case.

Reply:
@jbimbert  as mentioned https://issues.apache.org/jira/browse/DRILL-6603 @kkhatua tried your branch on the failed test and test was still failing because `stat.hasNonNullValue()` is missed.

Reply:
Dataset - http://drill-public.s3.amazonaws.com/data-shapes/wide-columns/5000/100000rows/parquet/widestrings.tar.gz

Reply:
@arina-ielchiieva thanks for the file. Running it, it happens that the parquet files have ST:[no stats for this column] and this is interpreted as "num_nulls=0, min/max undefined" in RangeExprEvaluator.visitTypedFieldExpr: 
if (columnStatistics != null) {
      return columnStatistics.getStatistics();
    } 
...
Which leads to the issue, since num_nulls is not defined in this case from the parquet rowgroup stats, but certainly not 0 (in this case, num_nulls > 0).

Reply:
Statistics class holds `numNulls` as long, so it will be 0 either defined or not. That's why we you need to add check with `stat.hasNonNullValue()`. Does it make sense?

Reply:
@arina-ielchiieva if `numNulls` is not defined, it is -1, not a 0.

Reply:
Yes. I have added hasNonNullValue() in createIsNullPredicate(). It solves the issue. All other tests OK.

Reply:
@arina-ielchiieva `num_nulls` must be -1 even for older files as well, otherwise, it is a bug in parquet library or drill that sets `num_nulls` incorrectly. If you have a concrete example where parquet library sets `num_nulls` to 0 where it should be -1, it is necessary to file JIRA for parquet. If drill sets `num_nulls` incorrectly while evaluating expressions or reading old parquet files, the bug needs to be fixed in drill. In the example you provided, should not statistics be empty (with `num_nulls` set to -1)?

AFAIK, it is not valid to use `hasNonNullValue()` in place for nulls check as min/max may not be defined when all values are nulls or when a comparator is not defined for a data type.

Reply:
@arina-ielchiieva checked all code and added hasNonNullValue() before  hasNoNulls.

Reply:
@vrozov, Drill test framework has such files (http://drill-public.s3.amazonaws.com/data-shapes/wide-columns/5000/100000rows/parquet/widestrings.tar.gz) and thus Drill producing the incorrect result, advanced tests are failing, please read details in Jira DRILL-6603. I have no idea where such files came from but maybe @kkhatua can give a clue.

Reply:
@vrozov it looks line stats did not have -1 be default all the time:
https://github.com/apache/parquet-mr/blob/apache-parquet-1.8.1/parquet-column/src/main/java/org/apache/parquet/column/statistics/Statistics.java
On;y starting from 1.10 things has changed: https://github.com/apache/parquet-mr/blob/apache-parquet-1.10.0/parquet-column/src/main/java/org/apache/parquet/column/statistics/Statistics.java

Reply:
@arina-ielchiieva It should not matter what was the "uninitialized" (or not set) value prior to 1.10. Drill now uses parquet library 1.10 where `isNumNullsSet()` is defined as **`num_nulls >= 0`**.

Reply:
I agree with your point but if Drill sticks to such rule, it will start reading incorrectly parquet files created by prev parquet versions which I believe this would be a great problem for the users who have files created earlier. Other option is to disable filter push down for files prior to 1.10 but this also will result in performance degradation. I suggest we at least try to support files created previously. Current implementation does so, if you have better ideas, please suggest. If you want to go by the road of ignoring previously created files stats, please start discussion on the mailing thread.

Reply:
@arina-ielchiieva I do not suggest that Drill does not support filter push down for parquet files created by any prior version. My suggestion is to see (debug) where `num_nulls` is not set correctly. In any case, it is not valid to rely on `hasNonNullValue()` as `hasNonNullValue()` may be `false` or `true` and `num_nulls` be 0 for both cases.

Note that `hasNoNulls()` is the inversion of `hasNulls()` condition, but the exact meaning is that either number of nulls is 0 or number of nulls is not set.

Reply:
@vrozov it turns out, I was mislead by the assumption that parquet files had incorrectly defined num of nulls. It turned out that actually it was Drill that read parquet stats and then re-created stats objects with num of nulls set to 0 instead of original -1.I'll prepare separate PR with the fix for it.
@jbimbert sorry for the confusion and thank you for your patience, please remove `stat.hasNonNullValue()` checks where you have added them.

Reply:
@arina-ielchiieva If I remove hasNonNullValue from createIsNullPredicate, the test mentioned inhttps://issues.apache.org/jira/browse/DRILL-6603 fails.
I can remove it in ParquetComparisonPredicate.checkNull(), but this function is used a lot... 
I can also remove it from ParquetIsPredicate.checkNull(), with the same comment.
For safety, until the bug is fixed, we maybe better keep it as is ?
(In the other places, it is used to ensure getMin/getMax is valid)

Reply:
I am working on PR to fix test failure mentioned in DRILL-6603. 
Please do not remove `stat.hasNonNullValue()` where it originally existed. Just remove where you have added the check when I asked you yesterday.

Reply:
Done. hasNonNullValue(0 removed in both checkNull() and createIsNullPredicate.